File: gestoteam-backend/src/main/java/com/gestoteam/service/TableConfigurationService.java
package com.gestoteam.service;

import com.gestoteam.dto.request.TableConfigurationRequest;
import com.gestoteam.dto.response.TableColumnConfigurationResponse;
import com.gestoteam.dto.response.TableConfigurationResponse;
import com.gestoteam.exception.GestoServiceException;
import com.gestoteam.model.TableConfiguration;
import com.gestoteam.model.TableColumnConfiguration;
import com.gestoteam.repository.TableConfigurationRepository;
import com.gestoteam.repository.UserRepository;
import lombok.extern.slf4j.Slf4j;
import org.modelmapper.ModelMapper;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@Service
@Slf4j
public class TableConfigurationService extends BaseService {

    private final TableConfigurationRepository tableConfigurationRepository;
    private final ModelMapper modelMapper;

    public TableConfigurationService(UserRepository userRepository, TableConfigurationRepository tableConfigurationRepository, ModelMapper modelMapper) {
        super(userRepository);
        this.tableConfigurationRepository = tableConfigurationRepository;
        this.modelMapper = modelMapper;
    }

    public TableConfigurationResponse getTableConfiguration(String tableName) {
        Long userId = getCurrentUserId();
        log.info("Obteniendo configuración de tabla '{}' para el usuario: {}", tableName, userId);

        try {
            TableConfiguration config = tableConfigurationRepository
                .findByUserIdAndTableNameOrderByUpdatedAtDesc(userId, tableName)
                .orElseGet(() -> createDefaultConfiguration(userId, tableName));
            
            log.info("Configuración de tabla '{}' obtenida para el usuario: {}", tableName, userId);
            return mapToResponse(config);
        } catch (Exception e) {
            log.error("Error al obtener configuración de tabla '{}' para el usuario: {}", tableName, userId, e);
            throw new GestoServiceException("Error al obtener la configuración de la tabla. Por favor, inténtelo más tarde.");
        }
    }

    @Transactional
    public TableConfigurationResponse saveTableConfiguration(TableConfigurationRequest request) {
        Long userId = getCurrentUserId();
        log.info("Guardando configuración de tabla '{}' para el usuario: {}", request.getTableName(), userId);

        try {
            // Buscar configuración existente
            TableConfiguration config = tableConfigurationRepository
                .findByUserIdAndTableNameOrderByUpdatedAtDesc(userId, request.getTableName())
                .orElseGet(() -> createDefaultConfiguration(userId, request.getTableName()));

            // Actualizar configuración general
            config.setPageSize(request.getPageSize());
            config.setDefaultSortKey(request.getDefaultSortKey());
            config.setDefaultSortOrder(TableConfiguration.SortOrder.valueOf(request.getDefaultSortOrder()));

            // Limpiar columnas existentes
            config.getColumnConfigurations().clear();

            // Crear y agregar nuevas columnas
            request.getColumnConfigurations().forEach(colReq -> {
                TableColumnConfiguration colConfig = TableColumnConfiguration.builder()
                    .tableConfiguration(config)
                    .columnKey(colReq.getColumnKey())
                    .columnLabel(colReq.getColumnLabel())
                    .visible(colReq.getVisible())
                    .sortable(colReq.getSortable())
                    .sortOrder(colReq.getSortOrder())
                    .width(colReq.getWidth())
                    .isDefault(false)
                    .build();
                
                config.addColumnConfiguration(colConfig);
            });

            // Guardar la configuración
            TableConfiguration savedConfig = tableConfigurationRepository.save(config);
            log.info("Configuración de tabla '{}' guardada para el usuario: {}", request.getTableName(), userId);
            
            return mapToResponse(savedConfig);
        } catch (Exception e) {
            log.error("Error al guardar configuración de tabla '{}' para el usuario: {}", request.getTableName(), userId, e);
            throw new GestoServiceException("Error al guardar la configuración de la tabla. Por favor, inténtelo más tarde.");
        }
    }

    public List<TableConfigurationResponse> getAllTableConfigurations() {
        Long userId = getCurrentUserId();
        log.info("Obteniendo todas las configuraciones de tabla para el usuario: {}", userId);

        try {
            List<TableConfiguration> configs = tableConfigurationRepository
                .findByUserIdOrderByTableNameAsc(userId);
            
            log.info("Se encontraron {} configuraciones de tabla para el usuario: {}", configs.size(), userId);
            return configs.stream()
                .map(this::mapToResponse)
                .collect(Collectors.toList());
        } catch (Exception e) {
            log.error("Error al obtener configuraciones de tabla para el usuario: {}", userId, e);
            throw new GestoServiceException("Error al obtener las configuraciones de tabla. Por favor, inténtelo más tarde.");
        }
    }

    @Transactional
    public void deleteTableConfiguration(String tableName) {
        Long userId = getCurrentUserId();
        log.info("Eliminando configuración de tabla '{}' para el usuario: {}", tableName, userId);

        try {
            tableConfigurationRepository.deleteByUserIdAndTableName(userId, tableName);
            log.info("Configuración de tabla '{}' eliminada para el usuario: {}", tableName, userId);
        } catch (Exception e) {
            log.error("Error al eliminar configuración de tabla '{}' para el usuario: {}", tableName, userId, e);
            throw new GestoServiceException("Error al eliminar la configuración de la tabla. Por favor, inténtelo más tarde.");
        }
    }

    private TableConfiguration createDefaultConfiguration(Long userId, String tableName) {
        log.info("Creando configuración por defecto para tabla '{}' y usuario: {}", tableName, userId);
        
        // Crear configuración con valores por defecto específicos para cada tabla
        TableConfiguration config = TableConfiguration.builder()
            .userId(userId)
            .tableName(tableName)
            .pageSize(10)
            .defaultSortKey(getDefaultSortKey(tableName))
            .defaultSortOrder(getDefaultSortOrder(tableName))
            .build();
        
        // Agregar columnas por defecto
        List<TableColumnConfiguration> defaultColumns = getDefaultColumnsForTable(tableName, config);
        defaultColumns.forEach(config::addColumnConfiguration);
        
        // Guardar la configuración
        TableConfiguration savedConfig = tableConfigurationRepository.save(config);
        
        log.info("Configuración por defecto creada para tabla '{}' y usuario: {}", tableName, userId);
        
        return savedConfig;
    }

    private String getDefaultSortKey(String tableName) {
        switch (tableName.toLowerCase()) {
            case "players":
                return "fullName";
            case "teams":
                return "name";
            case "matches":
                return "date";
            case "opponents":
                return "name";
            case "player-stats":
                return "starter";
            default:
                return "id";
        }
    }

    private TableConfiguration.SortOrder getDefaultSortOrder(String tableName) {
        switch (tableName.toLowerCase()) {
            case "matches":
            case "player-stats":
                return TableConfiguration.SortOrder.DESC;
            default:
                return TableConfiguration.SortOrder.ASC;
        }
    }

    private List<TableColumnConfiguration> getDefaultColumnsForTable(String tableName, TableConfiguration tableConfig) {
        List<TableColumnConfiguration> columns = new ArrayList<>();
        
        switch (tableName.toLowerCase()) {
            case "players":
                columns.add(createDefaultColumn("photoUrl", "Foto", true, false, 1, true, tableConfig));
                columns.add(createDefaultColumn("fullName", "Nombre", true, true, 2, true, tableConfig));
                columns.add(createDefaultColumn("position", "Posición", true, true, 3, true, tableConfig));
                columns.add(createDefaultColumn("foot", "Pie", true, true, 4, true, tableConfig));
                columns.add(createDefaultColumn("number", "Dorsal", true, true, 5, true, tableConfig));
                columns.add(createDefaultColumn("status", "Estado", true, true, 6, true, tableConfig));
                break;
            case "teams":
                columns.add(createDefaultColumn("name", "Nombre", true, true, 1, true, tableConfig));
                columns.add(createDefaultColumn("category", "Categoría", true, true, 2, true, tableConfig));
                columns.add(createDefaultColumn("division", "División", true, true, 3, true, tableConfig));
                columns.add(createDefaultColumn("location", "Ubicación", true, true, 4, true, tableConfig));
                break;
            case "matches":
                columns.add(createDefaultColumn("date", "Fecha", true, true, 1, true, tableConfig));
                columns.add(createDefaultColumn("opponent", "Rival", true, true, 2, true, tableConfig));
                columns.add(createDefaultColumn("homeAway", "Local/Visitante", true, true, 3, true, tableConfig));
                columns.add(createDefaultColumn("result", "Resultado", true, false, 4, true, tableConfig));
                break;
            case "opponents":
                columns.add(createDefaultColumn("name", "Nombre", true, true, 1, true, tableConfig));
                columns.add(createDefaultColumn("field", "Campo", true, true, 2, true, tableConfig));
                columns.add(createDefaultColumn("observations", "Observaciones", true, false, 3, true, tableConfig));
                break;
            case "player-stats":
                columns.add(createDefaultColumn("playerFullName", "Jugador", true, true, 1, true, tableConfig));
                columns.add(createDefaultColumn("minutesPlayed", "Minutos", true, true, 2, true, tableConfig));
                columns.add(createDefaultColumn("goals", "Goles", true, true, 3, true, tableConfig));
                columns.add(createDefaultColumn("assists", "Asist.", true, true, 4, true, tableConfig));
                columns.add(createDefaultColumn("cards", "Tarjetas", true, false, 5, true, tableConfig));
                break;
            default:
                columns.add(createDefaultColumn("id", "ID", true, true, 1, true, tableConfig));
                columns.add(createDefaultColumn("name", "Nombre", true, true, 2, true, tableConfig));
                break;
        }
        
        return columns;
    }

    private TableColumnConfiguration createDefaultColumn(String key, String label, boolean visible, boolean sortable, int order, boolean isDefault, TableConfiguration tableConfig) {
        return TableColumnConfiguration.builder()
            .tableConfiguration(tableConfig)
            .columnKey(key)
            .columnLabel(label)
            .visible(visible)
            .sortable(sortable)
            .sortOrder(order)
            .isDefault(isDefault)
            .build();
    }

    private TableConfigurationResponse mapToResponse(TableConfiguration config) {
        TableConfigurationResponse response = modelMapper.map(config, TableConfigurationResponse.class);
        
        // Mapear configuraciones de columnas
        List<TableColumnConfigurationResponse> columnResponses = config.getColumnConfigurations()
            .stream()
            .map(this::mapColumnToResponse)
            .collect(Collectors.toList());
        
        response.setColumnConfigurations(columnResponses);
        return response;
    }

    private TableColumnConfigurationResponse mapColumnToResponse(TableColumnConfiguration col) {
        return modelMapper.map(col, TableColumnConfigurationResponse.class);
    }
}

File: gestoteam-frontend/src/services/tableConfigurationService.js
import api from './api';

export const tableConfigurationService = {
  /**
   * Obtiene la configuración de una tabla específica
   * @param {string} tableName - Nombre de la tabla
   * @returns {Promise<Object>} Configuración de la tabla
   */
  async getTableConfiguration(tableName) {
    try {
      const response = await api.get(`/table-configurations/${tableName}`);
      return response.data;
    } catch (error) {
      console.error('Error al obtener configuración de tabla:', error);
      // Retornar configuración por defecto si hay error
      return this.getDefaultConfiguration(tableName);
    }
  },

  /**
   * Guarda la configuración de una tabla
   * @param {Object} configuration - Configuración a guardar
   * @returns {Promise<Object>} Configuración guardada
   */
  async saveTableConfiguration(configuration) {
    try {
      const response = await api.post('/table-configurations', configuration);
      return response.data;
    } catch (error) {
      console.error('Error al guardar configuración de tabla:', error);
      throw error;
    }
  },

  /**
   * Obtiene todas las configuraciones del usuario
   * @returns {Promise<Array>} Lista de configuraciones
   */
  async getAllTableConfigurations() {
    try {
      const response = await api.get('/table-configurations');
      return response.data;
    } catch (error) {
      console.error('Error al obtener configuraciones:', error);
      return [];
    }
  },

  /**
   * Elimina la configuración de una tabla
   * @param {string} tableName - Nombre de la tabla
   * @returns {Promise<void>}
   */
  async deleteTableConfiguration(tableName) {
    try {
      await api.delete(`/table-configurations/${tableName}`);
    } catch (error) {
      console.error('Error al eliminar configuración:', error);
      throw error;
    }
  },

  /**
   * Configuraciones por defecto para cada tipo de tabla
   * @param {string} tableName - Nombre de la tabla
   * @returns {Object} Configuración por defecto
   */
  getDefaultConfiguration(tableName) {
    const defaults = {
      players: {
        tableName: 'players',
        pageSize: 10,
        defaultSortKey: 'fullName',
        defaultSortOrder: 'ASC',
        columnConfigurations: [
          { columnKey: 'photoUrl', columnLabel: 'Foto', visible: true, sortable: false, sortOrder: 1, width: '80px' },
          { columnKey: 'fullName', columnLabel: 'Nombre', visible: true, sortable: true, sortOrder: 2, width: '200px' },
          { columnKey: 'position', columnLabel: 'Posición', visible: true, sortable: true, sortOrder: 3, width: '120px' },
          { columnKey: 'foot', columnLabel: 'Pie', visible: true, sortable: true, sortOrder: 4, width: '80px' },
          { columnKey: 'number', columnLabel: 'Dorsal', visible: true, sortable: true, sortOrder: 5, width: '80px' },
          { columnKey: 'status', columnLabel: 'Estado', visible: true, sortable: true, sortOrder: 6, width: '120px' }
        ]
      },
      teams: {
        tableName: 'teams',
        pageSize: 10,
        defaultSortKey: 'name',
        defaultSortOrder: 'ASC',
        columnConfigurations: [
          { columnKey: 'name', columnLabel: 'Nombre', visible: true, sortable: true, sortOrder: 1, width: '200px' },
          { columnKey: 'category', columnLabel: 'Categoría', visible: true, sortable: true, sortOrder: 2, width: '120px' },
          { columnKey: 'division', columnLabel: 'División', visible: true, sortable: true, sortOrder: 3, width: '120px' },
          { columnKey: 'location', columnLabel: 'Ubicación', visible: true, sortable: true, sortOrder: 4, width: '150px' }
        ]
      },
      matches: {
        tableName: 'matches',
        pageSize: 10,
        defaultSortKey: 'date',
        defaultSortOrder: 'DESC',
        columnConfigurations: [
          { columnKey: 'date', columnLabel: 'Fecha', visible: true, sortable: true, sortOrder: 1, width: '120px' },
          { columnKey: 'opponent', columnLabel: 'Rival', visible: true, sortable: true, sortOrder: 2, width: '150px' },
          { columnKey: 'homeAway', columnLabel: 'Local/Visitante', visible: true, sortable: true, sortOrder: 3, width: '120px' },
          { columnKey: 'result', columnLabel: 'Resultado', visible: true, sortable: false, sortOrder: 4, width: '100px' }
        ]
      },
      opponents: {
        tableName: 'opponents',
        pageSize: 10,
        defaultSortKey: 'name',
        defaultSortOrder: 'ASC',
        columnConfigurations: [
          { columnKey: 'name', columnLabel: 'Nombre', visible: true, sortable: true, sortOrder: 1, width: '200px' },
          { columnKey: 'field', columnLabel: 'Campo', visible: true, sortable: true, sortOrder: 2, width: '120px' },
          { columnKey: 'observations', columnLabel: 'Observaciones', visible: true, sortable: false, sortOrder: 3, width: '200px' }
        ]
      },
      'player-stats': {
        tableName: 'player-stats',
        pageSize: 10,
        defaultSortKey: 'starter',
        defaultSortOrder: 'DESC',
        columnConfigurations: [
          { columnKey: 'playerFullName', columnLabel: 'Jugador', visible: true, sortable: true, sortOrder: 1, width: '200px' },
          { columnKey: 'minutesPlayed', columnLabel: 'Minutos', visible: true, sortable: true, sortOrder: 2, width: '100px' },
          { columnKey: 'goals', columnLabel: 'Goles', visible: true, sortable: true, sortOrder: 3, width: '80px' },
          { columnKey: 'assists', columnLabel: 'Asist.', visible: true, sortable: true, sortOrder: 4, width: '80px' },
          { columnKey: 'cards', columnLabel: 'Tarjetas', visible: true, sortable: false, sortOrder: 5, width: '100px' }
        ]
      },
      trainings: {
        tableName: 'trainings',
        pageSize: 10,
        defaultSortKey: 'date',
        defaultSortOrder: 'DESC',
        columnConfigurations: [
          { columnKey: 'date', columnLabel: 'Fecha', visible: true, sortable: true, sortOrder: 1, width: '120px' },
          { columnKey: 'trainingType', columnLabel: 'Tipo', visible: true, sortable: true, sortOrder: 2, width: '120px' },
          { columnKey: 'location', columnLabel: 'Ubicación', visible: true, sortable: true, sortOrder: 3, width: '150px' },
          { columnKey: 'observations', columnLabel: 'Observaciones', visible: true, sortable: false, sortOrder: 4, width: '200px' }
        ]
      },
      attendance: {
        tableName: 'attendance',
        pageSize: 10,
        defaultSortKey: 'playerName',
        defaultSortOrder: 'ASC',
        columnConfigurations: [
          { columnKey: 'playerName', columnLabel: 'Jugador', visible: true, sortable: true, sortOrder: 1, width: '200px' },
          { columnKey: 'status', columnLabel: 'Estado', visible: true, sortable: true, sortOrder: 2, width: '120px' },
          { columnKey: 'notes', columnLabel: 'Notas', visible: true, sortable: false, sortOrder: 3, width: '200px' }
        ]
      },
      exercises: {
        tableName: 'exercises',
        pageSize: 10,
        defaultSortKey: 'title',
        defaultSortOrder: 'ASC',
        columnConfigurations: [
          { columnKey: 'title', columnLabel: 'Título', visible: true, sortable: true, sortOrder: 1, width: '200px' },
          { columnKey: 'category', columnLabel: 'Categoría', visible: true, sortable: true, sortOrder: 2, width: '120px' },
          { columnKey: 'description', columnLabel: 'Descripción', visible: true, sortable: false, sortOrder: 3, width: '250px' },
          { columnKey: 'tacticalObjectives', columnLabel: 'Obj. Tácticos', visible: true, sortable: false, sortOrder: 4, width: '150px' }
        ]
      }
    };

    return defaults[tableName] || defaults.teams;
  },

  /**
   * Aplica la configuración de columnas a una tabla existente
   * @param {Array} currentColumns - Columnas actuales de la tabla
   * @param {Object} userConfiguration - Configuración del usuario
   * @returns {Array} Columnas configuradas
   */
  applyColumnConfiguration(currentColumns, userConfiguration) {
    if (!userConfiguration || !userConfiguration.columnConfigurations) {
      return currentColumns.map(col => ({ ...col, visible: col.visible !== false }));
    }

    // Mapear configuración del usuario a las columnas actuales
    const configuredColumns = currentColumns.map(col => {
      const userColConfig = userConfiguration.columnConfigurations.find(
        config => config.columnKey === col.key
      );

      if (userColConfig) {
        return {
          ...col,
          visible: userColConfig.visible,
          sortable: userColConfig.sortable,
          width: userColConfig.width
        };
      } else {
        return { ...col, visible: col.visible !== false };
      }
    });

    // Ordenar columnas según la configuración del usuario
    return configuredColumns.sort((a, b) => {
      const configA = userConfiguration.columnConfigurations.find(
        config => config.columnKey === a.key
      );
      const configB = userConfiguration.columnConfigurations.find(
        config => config.columnKey === b.key
      );
      
      const orderA = configA ? configA.sortOrder : 999;
      const orderB = configB ? configB.sortOrder : 999;
      
      return orderA - orderB;
    });
  }
};
